jobbergate-core/jobbergate_core/auth/core.py

```diff
@@ -0,0 +1,316 @@
+"""
+Utilities for handling authentication in the Jobbergate system.
+"""
+import time
+from collections import namedtuple
+from dataclasses import dataclass, field
+from pathlib import Path
+from typing import Dict, cast
+
+import httpx
+from loguru import logger
+
+from .exceptions import AuthenticationError, TokenError
+from .token import Token, TokenType
+
+
+_LoginInformation = namedtuple(
+    "_LoginInformation",
+    ["verification_url", "wait_interval", "device_code"],
+)
+
+
+@dataclass
+class JobbergateAuth:
+    """
+    High-level class used to handle authentication in Jobbergate.
+
+    After an instance of this class is created, it can be used to authenticate requests
+    from both `requests`_  and `httpx`_ packages by passing it to the ``auth``
+    parameter on the request (see examples bellow).
+
+    It works just out of the box. Behind the scenes, this procedure calls the
+    :meth:`JobbergateAuth.acquire_tokens` method to load the available tokens from the cache directory, it
+    will try to refresh them if they are expired, or will provide an URL to the user to login on the system.
+
+    Notice all steps above are also available individually as public methods, allowing a fine control for
+    advanced users.
+
+    .. _requests: https://requests.readthedocs.io/en/latest/
+    .. _httpx: https://www.python-httpx.org/
+
+    Arguments:
+        cache_directory (Path): Directory to be used for the caching tokens.
+        login_domain (str): Domain used for the login.
+        login_audience (str): Audience of the login.
+        login_client_id (str): Client ID used for login.
+        tokens (Dict[TokenType, Token]): Dictionary holding the tokens needed for authentication.
+
+    Note:
+        Consult the values above with your system administrator.
+
+    Note:
+        This class can interoperate with the tokens generated by the ``jobbergate-cli`` package, as long
+        as they are stored in the same cache directory.
+
+    Examples:
+
+        The following example shows how to use the JobbergateAuth class to authenticate a request:
```

Comment:
```suggestion
        The following example shows how to use the :meth:`JobbergateAuth` class to authenticate a request:
```

Reply:
solved in [281a81e](https://github.com/omnivector-solutions/jobbergate/pull/228/commits/281a81ee6bffe72be79079ffa3676f75f87485a9)